File: tests/gpu_profile_json_test.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 Yunho Cho

#include <iostream>
#include <locale>

#include "../benchmarks/gpu_profile_json.h"

namespace {
struct CommaDecimal : std::numpunct<char> {
  char do_decimal_point() const override { return ','; }
  char do_thousands_sep() const override { return '_'; }
  std::string do_grouping() const override { return "\3"; }
};
}

int main(int argc, char** argv) {
  if (argc != 2) return 2;
  try {
    using namespace gjxl::benchmark;
    using namespace gjxl::gpu_profile_internal;
    std::locale::global(std::locale(std::locale::classic(), new CommaDecimal));
    GpuProfileJsonOptions options;
    options.scope = "fixture\n\"\\\b\f\r\t\x01";
    options.execution_path = "fixture-path\n";
    options.gpu_profiling_mode = GpuProfilingMode::kDispatch;
    options.gpu_aq = "fully-resident";
    options.ac_residual_inverse = "fused";
    options.butteraugli_target = 1.25f;
    options.samples = 1;
    GpuExecutionProfile profile;
    profile.mode = GpuProfilingMode::kDispatch;
    profile.capabilities = {true, true, true};
    profile.wall_stages.push_back({"host\n", GpuWallStageKind::kHost, 0, 12345});
    GpuStageProfile stage;
    stage.stage_id = "stage\"";
    stage.group_id = "group\\";
    stage.begin_timestamp = 12345;
    stage.end_timestamp = 23456;
    stage.gpu_nanoseconds = 11111;
    stage.timestamp_valid = true;
    stage.dispatches.push_back({"kernel\t", GpuDispatchKind::kThreadgroups,
                               {1, 2, 3}, {32, 4, 1}, 0, 12345, 23456, 11111, true});
    GpuSubmissionProfile submission;
    submission.submission_id = "submission\r";
    submission.command_buffer_gpu_nanoseconds = 11111;
    submission.stages.push_back(std::move(stage));
    GpuStageProfile empty_stage;
    empty_stage.stage_id = "empty-indirect";
    empty_stage.dispatches.push_back({"empty", GpuDispatchKind::kIndirectThreadgroups,
                                     {0, 1, 1}, {32, 1, 1}});
    submission.stages.push_back(std::move(empty_stage));
    profile.submissions.push_back(std::move(submission));
    RawGpuProfileWorkload workload{"workload\n", {1234, 9}, {}};
    workload.samples.push_back({0, std::move(profile)});
    std::vector<RawGpuProfileWorkload> workloads;
    workloads.push_back(std::move(workload));
    // Empty collections must remain valid JSON too.
    workloads.push_back({"empty", {1, 1}, {{0, {}}}});
    workloads.push_back({"no-samples", {1, 1}, {}});
    WriteGpuProfileSamples(std::filesystem::path(argv[1]), options, workloads);
    return 0;
  } catch (const std::exception& e) {
    std::cerr << e.what() << '\n';
    return 1;
  }
}

File: tests/gpu_profile_json_test.py
#!/usr/bin/env python3
"""Shared Metal/CUDA JSON writer: escaping, locale and atomic replacement."""

import json
from pathlib import Path
import subprocess
import sys
import tempfile


with tempfile.TemporaryDirectory(prefix="gjxl-profile-json-") as directory:
    root = Path(directory)
    destination = root / "nested" / "profile.json"
    for _ in range(2):
        result = subprocess.run([sys.argv[1], str(destination)], capture_output=True, text=True)
        assert result.returncode == 0, result.stderr
        document = json.loads(destination.read_text())
        assert document["scope"] == 'fixture\n"\\\b\f\r\t\x01'
        assert document["distance"] == 1.25
        assert document["schema_version"] == 4
        assert document["execution_path"] == "fixture-path\n"
        assert len(document["workloads"]) == 3
        workload = document["workloads"][0]
        assert workload["name"] == "workload\n" and workload["source_width"] == 1234
        sample = workload["samples"][0]
        assert sample["wall_stages"][0]["wall_nanoseconds"] == 12345
        submission = sample["submissions"][0]
        assert submission["submission_id"] == "submission\r"
        stage = submission["stages"][0]
        assert stage["stage_id"] == 'stage"' and stage["group_id"] == "group\\"
        assert stage["dispatches"][0]["kernel_id"] == "kernel\t"
        assert stage["dispatches"][0]["gpu_nanoseconds"] == 11111
        assert stage["timestamp_valid"] is True
        assert stage["dispatches"][0]["timestamp_valid"] is True
        empty = submission["stages"][1]
        assert empty["timestamp_valid"] is False and empty["gpu_nanoseconds"] == 0
        assert empty["dispatches"][0]["kind"] == "indirect_threadgroups"
        assert empty["dispatches"][0]["grid"] == [0, 1, 1]
        assert empty["dispatches"][0]["timestamp_valid"] is False
        assert document["workloads"][1]["samples"][0]["submissions"] == []
        assert document["workloads"][2]["samples"] == []
        assert not list(root.rglob("*.tmp-*"))
    blocked = root / "blocked.json"
    blocked.mkdir()
    (blocked / "sentinel").write_text("keep")
    result = subprocess.run([sys.argv[1], str(blocked)], capture_output=True, text=True)
    assert result.returncode == 1
    assert (blocked / "sentinel").read_text() == "keep"
    assert not list(root.rglob("*.tmp-*"))
    sentinel = destination.read_bytes()
    result = subprocess.run([sys.argv[1], str(destination / "child")], capture_output=True, text=True)
    assert result.returncode == 1
    assert destination.read_bytes() == sentinel
    assert not list(root.rglob("*.tmp-*"))
print("Shared GPU-profile JSON escaping, locale, replacement and failure cleanup passed")
